File: LV 2/행렬 테두리 회전하기.py
def rotate(matrix, x1, y1, x2, y2):
    minvalue = 999999999
    last = matrix[x1][y1]
    
    for i in range(x1, x2):
        matrix[i][y1] = matrix[i+1][y1]
        minvalue = min(minvalue, matrix[i+1][y1])
        
    for j in range(y1, y2):
        matrix[x2][j] = matrix[x2][j+1]
        minvalue = min(minvalue, matrix[x2][j+1])
    
    for i in range(x2-1, x1-1, -1):
        matrix[i+1][y2] = matrix[i][y2]
        minvalue = min(minvalue, matrix[i][y2])
        
    for j in range(y2-1, y1, -1):
        matrix[x1][j+1] = matrix[x1][j]
        minvalue = min(minvalue, matrix[x1][j])
        
    matrix[x1][y1+1] = last
    minvalue = min(minvalue, last)
    return minvalue
    
def solution(rows, columns, queries):
    answer = []
    
    matrix = [[0] * (columns + 1) for _ in range(rows + 1)]
    for i in range(1, rows+1):
        for j in range(1, columns+1):
            matrix[i][j] = columns*(i-1)+j
    
    for query in queries:
        answer.append(rotate(matrix, query[0], query[1], query[2], query[3]))
    
    return answer

File: LV 2/행렬 테두리 회전하기.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public int[][] makeArray(int rows, int columns) {
        int[][] arr = new int[rows][columns];
        int idx = 1;
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < columns; j++) {
                arr[i][j] = idx++;
            }
        }

        return arr;
    }

    public int movement(int[][] arr, int[] query) {
        int startRow = query[0] - 1, startCol = query[1] - 1;
        int finishRow = query[2] - 1, finishCol = query[3] - 1;

        int lastValue = arr[startRow][startCol];
        int minValue = 10005;
        // 왼쪽
        for (int i = startRow; i < finishRow; i++) {
            arr[i][startCol] = arr[i+1][startCol];
            minValue = Math.min(minValue, arr[i][startCol]);
        }

        // 아래
        for (int i = startCol; i < finishCol; i++) {
            arr[finishRow][i] = arr[finishRow][i+1];
            minValue = Math.min(minValue, arr[finishRow][i]);
        }

        // 오른쪽
        for (int i = finishRow; i > startRow; i--) {
            arr[i][finishCol] = arr[i-1][finishCol];
            minValue = Math.min(minValue, arr[i][finishCol]);
        }

        // 위
        for (int i = finishCol; i > startCol; i--) {
            arr[startRow][i] = arr[startRow][i-1];
            minValue = Math.min(minValue, arr[startRow][i-1]);
        }
        
        arr[startRow][startCol+1] = lastValue;
        return Math.min(minValue, lastValue);
    }

    public int[] solution(int rows, int columns, int[][] queries) {
        List<Integer> list = new ArrayList<>();
        int[][] arr = makeArray(rows, columns);
        for (int[] query : queries) {
            int minValue = movement(arr, query);
            list.add(minValue);
        }
        return list.stream().mapToInt(i->i).toArray();
    }
}
